gptfixer/gpt.c

```diff
@@ -0,0 +1,914 @@
+    err(1, "Cannot open %s", dev);
+  }
+  int signed_sector_size;
+  if (!bool_ioctl(fd, BLKSSZGET, &signed_sector_size)) {
+    err(1, "ioctl(%d, BLKSSZGET, %p)", fd, &signed_sector_size);
+  }
+
+  const uint32_t sector_size =
+      (uint32_t)signed_sector_size; // size in bytes of a sector
+
+  if (sector_size != 512 && sector_size != 4096) {
+    errx(1,
+         "block size is 0x%" PRIX32
+         " is neither 512 or 4096 and we don't support that",
+         sector_size);
+  }
+
+  uint64_t size; // size in bytes
+  if (!bool_ioctl(fd, BLKGETSIZE64, &size)) {
+    err(1, "ioctl(%d, BLKGETSIZE64, %p)", fd, &size);
+  }
+
+  if ((off_t)size < 0 || (uint64_t)(off_t)size != size) {
+    errx(1, "size 0x%" PRIx64 " does not fit in off_t", size);
+  }
+
+  if (size % sector_size != 0) {
+    errx(1, "size 0x%" PRIx64 " is not multiple of sector size 0x%" PRIx32,
+         size, sector_size);
+  }
+  const uint64_t sectors = size / sector_size; // number of sectors
+  if (sectors < 6) {
+    errx(1, "Device has too few sectors for GPT: need 6, found %" PRIu64,
+         sectors);
+  }
+
+  if (fsync(fd)) {
+    err(1, "fsync(\"%s\")", argv[1]);
+  }
+  {
+    char *buffer = aligned_alloc(sector_size, sector_size);
+    if (pread(fd, buffer, sector_size, 0) != sector_size) {
+      err(1, "Failed reading MBR from block device or short read");
+    }
+    static_assert(sizeof(struct Mbr) <= 512, "Bad MBR size");
+    check_mbr((struct Mbr *)buffer, sectors);
+    vlog("Sector size is %" PRIu32 ", device size is 0x%" PRIx64 ", 0x%" PRIx64
+         " sectors",
+         sector_size, size, sectors);
+    free(buffer);
+  }
+
+  struct GPT *gpt = read_and_check_gpt(fd, 1, sector_size, sectors);
+  if (gpt == NULL) {
+    gpt = read_and_check_gpt(fd, sectors - 1, sector_size, sectors);
+  }
+  bool try_changing = gpt == NULL;
+  if (try_changing) {
+    uint32_t new_sector_size = (sector_size == 4096) ? 512 : 4096;
+    uint64_t new_sectors =
+        (sector_size == 4096) ? (sectors << 3) : (sectors >> 3);
+    bool used_primary = true;
+    gpt = read_and_check_gpt(fd, 1, new_sector_size, new_sectors);
+    if (gpt == NULL) {
+      used_primary = false;
+      gpt =
+          read_and_check_gpt(fd, new_sectors - 1, new_sector_size, new_sectors);
+    }
+    if (gpt == NULL) {
+      errx(1, "Cannot find valid GPT");
+    }
+    warnx("Found GPT with different sector size, altering");
+    // "First, do no harm": always start by overwriting the GPT we
+    // did *not* use, so there is always at least one good GPT!
+    struct GPT *new_gpt;
+    uint64_t to_write = 0;
+    if (new_sector_size < sector_size) {
+      new_gpt = gpt_grow_sectors(used_primary ? sectors - 1 : 1, sector_size,
+                                 new_sector_size, sectors, gpt, &to_write);
+    } else {
+      new_gpt = gpt_shrink_sectors(used_primary ? sectors - 1 : 1, sector_size,
+                                   new_sector_size, sectors, gpt, &to_write);
```

Comment:
The backup GPT is only looked at if the primary GPT is invalid.  If the primary GPT is valid, the backup GPT is written before the primary GPT is.  Since a GPT cannot be written atomically, this requires that the backup GPT must be written at the end of the disk, so that if there is a problem (such as power failure) either the backup GPT is now valid (at the new sector size) or the old primary GPT (with the old sector size) is still present.  Furthermore, writing a GPT will clobber any other GPTs at the same end of the disk, so if the primary GPT was used the backup GPT must be written first and visa versa.  Therefore, adjusting the GPT size to the actual disk size is unavoidable, unless I am missing something.

Reply:
> The backup GPT is only looked at if the primary GPT is invalid.

That's not true. At the very least, fdisk will complain if backup GPT doesn't match the primary GPT even if primary GPT is valid.

Anyway, I'm not going to argue about this too much, it's a very rare case anyway (resizing the disk and changing sector size at the same time). Just ensure you don't resize GPT if you don't change sector size (I think this case is okay currently).

Reply:
This may be a corner case, but it is still a case I want to handle correctly.  Can you explain, in detail, how this could go wrong?

My understanding is:

- Tools must create a primary and backup GPT.
- Tools must accept a primary GPT that is valid, even if the backup GPT is invalid.
- Tools must use the primary GPT if it is valid.
- If the primary GPT is valid but the backup GPT is invalid or visa versa, tools must not reject the GPT altogether but are allowed to issue a warning.

Do real-world tools behave differently than this?

Reply:
> Take location of the other GPT from the header, instead of hardcoding to `sectors - 1` - otherwise it will explode on first boot after resizing disk (if that happens to be also first boot on a disk with different sector size). Same in few other locations below.

Do you mean that if the primary GPT _header_ is valid, but the primary GPT _entries_ are invalid, the backup GPT _header_ should be read from `AlternateLBA` in the primary header?

Most of the places where `sectors - 1` is written refer to either:

1. Where to _write_ the backup GPT.  The backup GPT is meant as a backup if the primary GPT is bad, so it needs to be written at the well-known location regardless of where the primary GPT is.
2. Where to find the backup GPT if the primary GPT is bad.  In this case (such as if e.g. the primary header CRC32 checksum is invalid), there is no point (that I can think of) in using the primary GPT header’s `AlternateLBA` at all, as it’s presumably garbage.

Reply:
> (don't try to adjust GPT size to the actual disk size - leave that to the OS)

Ah, `qubes-rootfs-resize.service`.  That could also use this tool to find the partition to resize, instead of hardcoding 3 (but if you want to hardcode 3, that’s fine too).